docs: document trusted boot (#2612)
spectrocloud/librarium: docs/docs-content/clusters/edge/trusted-boot/deployment-day2/upgrade-cluster.md

title: "Update Cluster with Trusted Boot"
description: "Learn about Trusted Boot."
hide_table_of_contents: false
sidebar_position: 60
tags: ["edge"]
---

If you need to make changes to a deployed Edge cluster with Trusted Boot enabled without touching the OS or the
Kubernetes layer, you can follow the regular cluster upgrade process. Refer to
[Update a Cluster](/docs/docs-content/clusters/cluster-management/cluster-updates.md) for more information. However, if
the change you are making involves the Operating System (OS) layer and the Kubernetes layer, you will need to build a
new provider image with the same Trusted Boot keys you used to build the Edge Installer ISO.

This page guides you through the process of making an update to a cluster that involves the Operating System (OS) or
Kubernetes layer of the cluster.

## Prerequisites

- You have built an Edge Installer ISO with Trusted Boot enabled and used the ISO to deploy a cluster. The Edge hosts
  have keys enrolled.

- You have access to the original keys that were used to build the original ISO and provider images. Refer to
  [Build Edge Installer ISO with Trusted Boot](../edgeforge/build-trusted-iso.md) for details.

> - The machine used to build the provider imaged must meet the minimum hardware requirements:

Reviewer: ```suggestion
- The machine used to build the provider image must meet the minimum hardware requirements:
```

  - 4 CPUs
  - 32 GB RAM
  - 100 GB storage

:::warning

Ensure that the keys in the **secure-boot** folder contains all the keys that were used to produce the ISO, including

> the keys in each subdirectory. If you are missing any of the keys, your Edge device will either not be able to boot, or

Reviewer: ```suggestion
the keys in each subdirectory. If you are missing any of the keys, your Edge host will either not be able to boot, or
```

---
sidebar_label: "Update Cluster"
title: "Update Cluster with Trusted Boot"
description: "Learn about Trusted Boot."
hide_table_of_contents: false
sidebar_position: 60
tags: ["edge"]
---

> If you need to make changes to a deployed Edge cluster with Trusted Boot enabled without touching the OS or the

Reviewer 2: ```suggestion
If you need to make changes to a deployed Edge cluster with Trusted Boot enabled without modifying the Operating System (OS) or the
```

Replaced "touching" to the more formal form better fitted in technical documentation.

Kubernetes layer, you can follow the regular cluster upgrade process. Refer to

> [Update a Cluster](/docs/docs-content/clusters/cluster-management/cluster-updates.md) for more information. However, if

Reviewer 2: Let's use relative URLs please.

the change you are making involves the Operating System (OS) layer and the Kubernetes layer, you will need to build a
new provider image with the same Trusted Boot keys you used to build the Edge Installer ISO.

This page guides you through the process of making an update to a cluster that involves the Operating System (OS) or
Kubernetes layer of the cluster.

## Limitation

- Palette does not stop the upgrade if the provider image that is not recognized by secure boot. If you upgrade to using
  an unrecognized image, you will receive a secure boot violation. This will require you to choose the recovery option

>   from the GRand Unified Bootloader (GRUB) menu to return the Edge host to boot using the previous image.

Reviewer 2: ```suggestion
  from the GRand Unified Bootloader (GRUB) Menu to return the Edge host to boot using the previous image.
```

## Prerequisites

> 

Reviewer 2: Seems like having CanvOS download should be a prereq, along with git.

   | `OS_DISTRIBUTION`  | OS distribution.                                                                                      | `ubuntu`, `opensuse-leap`, `rhel`.         |
   | `OS_VERSION`       | OS version. This applies to Ubuntu only.                                                              | `20`, `22`.                                |

4. Issue the following command to build the provider image.

   ```shell
   ./earthly.sh +build-provider-images
   ```

   ```hideClipboard bash
   # Output condensed for readability
   ===================== Earthly Build SUCCESS =====================
   Share your logs with an Earthly account (experimental)! Register for one at https://ci.earthly.dev.
   ```

   You should also receive an output from the build command that contains the OS layer of your cluster profile. Copy and
   save it during the later steps.

5. Issue the following command to push the image to the repository you specified in **.arg**. Replace `IMAGE-TAG` with
   the tag of the provider image you just built.

   ```shell
   docker push [IMAGE-TAG]
   ```

> 

Reviewer 2: We are missing the step where the user pushes the images to a registry.

  - 32 GB RAM
  - 100 GB storage

> - [Git](https://cli.github.com/manual/installation). You can ensure git installation by issuing the git --version

Reviewer 2: ```suggestion
- [Git](https://www.git-scm.com/downloads). You can ensure git installation by issuing the git --version
```

  command.

## Instructions

1. If you have access to the original **CanvOS** folder where they provider images were made, and all the necessary keys
   are available to you in their original folders, skip to step 5.

2. Check out the [CanvOS](https://github.com/spectrocloud/CanvOS) GitHub repository containing the starter code.

   ```bash
   git clone https://github.com/spectrocloud/CanvOS.git
   ```

3. Change to the **CanvOS/** directory.

   ```bash
   cd CanvOS
   ```

4. View the available [git tag](https://github.com/spectrocloud/CanvOS/tags).

   ```bash
   git tag
   ```

5. Check out the newest available tag. This guide uses the tag **v4.4.0** as an example.

   ```shell
   git checkout v4.4.0
   ```

6. Move all keys into folders where they are expected. Ensure all keys that were used to build the ISO are present in
   the **secure-boot** folder. The following files and folder structure are expected with the following file
   permissions.

>    ```

Reviewer 2: Let's seperate the command and the output. That way we can use hideClipboard on the output :)

   You should also receive an output from the build command that contains the OS layer of your cluster profile. Copy and
   save it during the later steps.

> 9. Issue the following command to push the image to the repository you specified in **.arg**. Replace `IMAGE-TAG` with

Reviewer 2: One minor note to callout to the reader. Depending on what repository they used, they may have to log in to their repository through the repo's preferred authentication method.
